### PR TITLE
Fix nextPowerOfTwo in math

### PR DESCRIPTION
**Pull request by koalazen**
nextPowerOfTwo doesn't work as expected for the smallest integer. 

echo 1 shl 63, " next power of two : ", nextPowerOfTwo(1 shl 63)
-9223372036854775808 next power of two : -9223372036854775807

This patch fixes it by incrementing this smallest integer first (and possibly other negative integers but it doesn't change the result).

### CODE REVIEW

File: lib/pure/math.nim [hunk 1]
```diff
@@ -90,7 +90,19 @@ proc isPowerOfTwo*(x: int): bool {.noSideEffect.} =
 proc nextPowerOfTwo*(x: int): int {.noSideEffect.} =
   ## returns `x` rounded up to the nearest power of two.
   ## Zero and negative numbers get rounded up to 1.
-  result = x - 1 
+
+  # Increment x if x is the smallest integer.
+  result = x
+  when defined(cpu64):
```
Comment (Varriount) [inline lib/pure/math.nim]: Are all these conditionals really needed? Also, wouldn't it be better to test on the type, rather than the size?

**Comment by Araq** [comment]
Testcases missing.

File: lib/pure/math.nim (hunk 1 above)
Comment (koalazen) [inline lib/pure/math.nim]: The type is int. I could simply do if(x <0) return 1 but I tried to stick to the approach of this function. It seems faster like that.

**Comment by koalazen** [comment]
Hmm agree that this needs tests. Where should I add them ? Are there some guidelines ?

**Comment by reactormonk** [comment]
Yup, you found the correct location.

File: lib/pure/math.nim [hunk 2]
```diff
@@ -90,7 +90,7 @@ proc isPowerOfTwo*(x: int): bool {.noSideEffect.} =
 proc nextPowerOfTwo*(x: int): int {.noSideEffect.} =
   ## returns `x` rounded up to the nearest power of two.
   ## Zero and negative numbers get rounded up to 1.
-  result = x - 1 
+  result = x + ord(x<0) - 1
```
Comment (Araq) [inline lib/pure/math.nim]: Meh, who cares about nextPowerOfTwo for negative numbers? What's the use case? Alternatively 'nextPowerOfTwo' should take 'x: Natural'.

**Comment by koalazen** [comment]
Ok so I changed it to use Natural instead. Now for naturals above the biggest power of two the function returns 1. So the function loops over the powers of two.

File: lib/pure/math.nim [hunk 3]
```diff
@@ -100,7 +100,12 @@ proc nextPowerOfTwo*(x: int): int {.noSideEffect.} =
   result = result or (result shr 4)
   result = result or (result shr 2)
   result = result or (result shr 1)
-  result += 1 + ord(x<=0)
+  result += 1 
+  # Handle the case where x is higher than the greatest power of two.
```
Comment (Araq) [inline lib/pure/math.nim]: meh, that's more complex than your previous solution. If it's that complex, better leave it working on ``int``?

**Comment by Araq** [comment]
Well I don't like this. I guess this proc should not be in ``math`` but perhaps in ``bitops``. It's supposed to be as fast&small as possible for hashing stuff. For this it cannot overflow or underflow since you run out of memory first.

**Pull request closed without merging**